{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import joblib\n",
    "from tqdm.notebook import tqdm\n",
    "import warnings\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pravki\n",
    "#не, 1) у тебя было /path/ это не относительный, \n",
    "#а 2) в коде у тебя resources, а в архиве у тебя сразу Files/Gpcrdb_files/...\n",
    "#задаешь workding_dir, например \"./\", а потом все остальные через os.path.join(workding_dir, ...)\n",
    "\n",
    "\n",
    "\n",
    "#open(dir_path+\"{}_structure_seq.fasta\".format(pdb_id) - везде заменяй на os.path.join\n",
    "\n",
    "#в GPCRdb_mapping_for_sequence не нужно два раза считывать файл\n",
    "#Видимо, мой прошлый коммент про считывания относился к этому\n",
    "#if \"ATOM\" in line:\n",
    "#                seq_num.append(int(line[22:31]))\n",
    "#Считай правильно по формату\n",
    "#os.system(\"mafft --addfull {} --mapout {} > ns_aligned_to_db.fasta\".format(dir_path+pdb_id+\"_structure_seq.fasta\", path_to_gpcrdb_files+\"GPCRdb_alignment_honly.fasta\"))\n",
    "\n",
    "#ресурсные файлы нужно вынести в отдельный cell и обозвать через переменные,\n",
    "#например fn_gpcrdb_aln_honly = \"GPCRdb_alignment_honly.fasta\"\n",
    "#map_seq[\"GPCRdb_numeration\"] = ... - плохо читается\n",
    "#[\"1x50\", \"2x50\", \"3x50\", \"4x50\", \"5x50\", \"6x50\", \"7x50\", \"8x50\"] - вынести в ресурсную переменную\n",
    "\n",
    "\n",
    "#...\"1x50\"][\"# letter\"].tolist()[0] != \"N\" - лучше составить словарь {'1x50':'N'} и проверять через него\n",
    "#тогда для 2.50 можно сделать в одну команду ({'2.50':'DN'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "working_dir = \"./\"\n",
    "dir_path = os.path.join(working_dir,\"test\")# path to working directory\n",
    "out_path = os.path.join(working_dir,\"test\")# path to output directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "in case of singe-pdb file analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_id = \"identifier1\"#file identifier\n",
    "PDBFile = os.path.join(working_dir,\"Files/examples/Single_pdbs/2YDV_frame_22_act.pdb\")#path to single pdb file\n",
    "path_to_gpcrdb_files = os.path.join(working_dir,\"Files/resources/Gpcrdb_files/\")\n",
    "one_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/onemodal_features_dist_stats.csv\")\n",
    "bi_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/bimodal_features_dist_stats.csv\")\n",
    "model = os.path.join(working_dir,\"Files/resources/RF_classifier_good_model_19.joblib\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "in case of MD trajectory analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_id = \"identifier1\"#file identifier\n",
    "frames_dir_path = os.path.join(working_dir,\"Files/examples/35_10386_apoform_3REY\")#path to trajectory pdb files\n",
    "path_to_gpcrdb_files = os.path.join(working_dir,\"Files/resources/Gpcrdb_files/\")\n",
    "one_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/onemodal_features_dist_stats.csv\")\n",
    "bi_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/bimodal_features_dist_stats.csv\")\n",
    "model = os.path.join(working_dir,\"Files/resources/RF_classifier_good_model_19.joblib\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "resource files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_seq_aligned_to_GPCRdb_seq_database = \"ns_aligned_to_db.fasta\"\n",
    "canonical_residues = [\"1x50\", \"2x50\", \"3x50\", \"4x50\", \"5x50\", \"6x50\", \"7x50\", \"8x50\"]\n",
    "#PP: check aa, e.g. for 2.50 should be 'DN'\n",
    "canonical_residues_dict = {\"1x50\":\"N\", \"2x50\":\"D\", \"3x50\":\"R\", \"4x50\":\"W\", \"5x50\":\"P\", \"6x50\":\"P\", \"7x50\":\"P\", \"8x50\":\"F\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GPCRdb_mapping_for_sequence(pdb_id, PDBFile, path_to_gpcrdb_files, dir_path, out_path, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict):\n",
    "    seq_num = []\n",
    "    \n",
    "    #read initial sequence\n",
    "    with open(PDBFile, 'r') as pdb_file:\n",
    "        with open(os.path.join(dir_path,\"{}_structure_seq.fasta\".format(pdb_id)), \"w\") as m:\n",
    "            for record in SeqIO.parse(pdb_file, 'pdb-atom'):\n",
    "                print('>' + pdb_id, file = m)\n",
    "                print(record.seq, file = m)\n",
    "                \n",
    "    #read initial numeration\n",
    "    with open(PDBFile, 'r') as pdb_file:\n",
    "        for line in pdb_file:\n",
    "            if \"ATOM\" in line:\n",
    "                seq_num.append(int(line[22:31])) #PP: you need only res number here, so use only res number fields. Use 'replace' to remove spaces, if needed.\n",
    "    seq_num = list(set(seq_num))\n",
    "    \n",
    "    #align initial sequence to GPCRdb alignment\n",
    "    \n",
    "    #PP: define all magic constants (mc) (e.g. \"GPCRdb_alignment_honly.fasta\") in the corresponding upper cell \n",
    "    os.system(\"mafft --addfull {} --mapout {} > {}\".format(os.path.join(dir_path,(pdb_id+\"_structure_seq.fasta\")), os.path.join(path_to_gpcrdb_files,\"GPCRdb_alignment_honly.fasta\"), new_seq_aligned_to_GPCRdb_seq_database))\n",
    "    \n",
    "    #get initial position numeration according to the alogned sequence\n",
    "    map_seq = pd.read_csv(os.path.join(dir_path,(pdb_id+\"_structure_seq.fasta.map\")), skiprows=1)\n",
    "    map_seq[\"original position from pdb file\"] = seq_num\n",
    "    #PP: mc; split into several commands\n",
    "    map_seq[\"GPCRdb_numeration\"] = [pd.read_csv(os.path.join(path_to_gpcrdb_files,\"gpcrdb_numbers_honly.csv\"))[\"Numeration\"].tolist()[int(f)-1] if f  != ' -' else \" -\" for f in map_seq[\" position in the reference alignment\"].tolist()]\n",
    "    map_seq[\"original_sequence\"] = map_seq[\"# letter\"] +map_seq[\"original position from pdb file\"].astype(\"str\")\n",
    "    map_seq.to_csv(os.path.join(out_path,\"{}_mapping.csv\".format(pdb_id)))\n",
    "    \n",
    "    #test for invalid mapping or mutant(N1.50: 98%, D2.50: 90%, R3.50: 95%, W4.50: 97%, P5.50: 78%, P6.50: 99%, P7.50: 88% F8x50)https://pubmed.ncbi.nlm.nih.gov/24304901/\n",
    "    #conserved aa residues check\n",
    "    \n",
    "    check_map_df = map_seq[map_seq[\"GPCRdb_numeration\"].isin(canonical_residues)].copy()\n",
    "    \n",
    "    for aa in canonical_residues_dict.keys():\n",
    "        # PP aa is a letter, e.g. D, dict value is a string, e.g. DN, so use 'isin', rather than !=\n",
    "        if check_map_df[check_map_df[\"GPCRdb_numeration\"] == aa][\"# letter\"].tolist()[0] != canonical_residues_dict[aa]:\n",
    "            warnings.warn(\"The amino acid residue at position {} seems not to be a canonical one ({}). Please re-check the mapping results!\".format(aa, canonical_residues_dict[aa])) \n",
    "#    if check_map_df[check_map_df[\"GPCRdb_numeration\"] == \"1x50\"][\"# letter\"].tolist()[0] != \"N\":\n",
    "#        warnings.warn(\"The amino acid residue at position 1x50 seems not to be a canonical one (N). Please re-check the mapping results!\")\n",
    "#    if check_map_df[check_map_df[\"GPCRdb_numeration\"] == \"2x50\"][\"# letter\"].tolist()[0] != \"D\":\n",
    "#        warnings.warn(\"The amino acid residue at position 2x50 seems not to be a canonical one (D). Please re-check the mapping results!\")\n",
    "#    if check_map_df[check_map_df[\"GPCRdb_numeration\"] == \"3x50\"][\"# letter\"].tolist()[0] != \"R\":\n",
    "#        warnings.warn(\"The amino acid residue at position 3x50 seems not to be a canonical one (R). Please re-check the mapping results!\")\n",
    "#    if check_map_df[check_map_df[\"GPCRdb_numeration\"] == \"4x50\"][\"# letter\"].tolist()[0] != \"W\":\n",
    "#        warnings.warn(\"The amino acid residue at position 4x50 seems not to be a canonical one (W). Please re-check the mapping results!\")\n",
    "#    if check_map_df[check_map_df[\"GPCRdb_numeration\"] == \"5x50\"][\"# letter\"].tolist()[0] != \"P\":\n",
    "#        warnings.warn(\"The amino acid residue at position 5x50 seems not to be a canonical one (P). Please re-check the mapping results!\")\n",
    "#    if check_map_df[check_map_df[\"GPCRdb_numeration\"] == \"6x50\"][\"# letter\"].tolist()[0] != \"P\":\n",
    "#        warnings.warn(\"The amino acid residue at position 6x50 seems not to be a canonical one (P). Please re-check the mapping results!\")\n",
    "#    if check_map_df[check_map_df[\"GPCRdb_numeration\"] == \"7x50\"][\"# letter\"].tolist()[0] != \"P\":\n",
    "#        warnings.warn(\"The amino acid residue at position 7x50 seems not to be a canonical one (P). Please re-check the mapping results!\")\n",
    "#    if check_map_df[check_map_df[\"GPCRdb_numeration\"] == \"8x50\"][\"# letter\"].tolist()[0] != \"F\":\n",
    "#        warnings.warn(\"The amino acid residue at position 8x50 seems not to be a canonical one (F). Please re-check the mapping results!\")\n",
    "    \n",
    "    print(os.path.join(out_path, \"{}_mapping.csv\".format(pdb_id)))\n",
    "    \n",
    "    return map_seq\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_dist_feature_modif_no_c_id(PDBFile, map_df, pdb_id, one_mod_df, bi_mod_df):\n",
    "    \n",
    "    #PP: mc\n",
    "    d = {'CYS': 'C', 'ASP': 'D', 'SER': 'S', 'GLN': 'Q', 'LYS': 'K',\n",
    "     'ILE': 'I', 'PRO': 'P', 'THR': 'T', 'PHE': 'F', 'ASN': 'N', \n",
    "     'GLY': 'G', 'HIS': 'H', 'LEU': 'L', 'ARG': 'R', 'TRP': 'W', \n",
    "     'ALA': 'A', 'VAL':'V', 'GLU': 'E', 'TYR': 'Y', 'MET': 'M'}#, 'HIS': 'X'}\n",
    "    inv_d = {v: k for k, v in d.items()}\n",
    "    mapping = map_df\n",
    "    map_dict = {}\n",
    "    for i in range(len(mapping[\"original_sequence\"].dropna())):\n",
    "        if mapping[\"GPCRdb_numeration\"].tolist()[i] == \"-\":\n",
    "            continue\n",
    "        map_dict[str(mapping[\"original_sequence\"].dropna().tolist()[i])] = mapping[\"GPCRdb_numeration\"].tolist()[i]\n",
    "    inv_map_dict = {v: k for k, v in map_dict.items()}\n",
    "    i=0\n",
    "    #PP: mc\n",
    "    res_contact_list = [\"1x49:7x50\", \"1x53:7x53\", \"1x53:7x54\", \"2x37:2x40\", \"2x42:4x45\", \"2x43:7x53\", \"2x45:4x50\", \"2x46:2x50\", \"2x50:3x39\", \"2x50:7x49\", \"2x57:7x42\", \"3x40:6x48\", \"3x43:6x40\", \"3x43:6x41\", \"3x43:7x49\", \"3x43:7x53\", \"3x46:6x37\", \"3x46:7x53\", \"3x46:3x50\",\"3x49:3x50\", \"3x50:3x53\", \"3x50:6x37\", \"3x50:7x53\", \"3x51:5x57\", \"5x51:6x44\",\"5x55:6x41\", \"5x58:6x40\", \"5x62:6x37\", \"6x40:7x49\", \"6x44:6x48\",\"6x44:7x45\",\"6x48:7x45\", \"7x45:7x49\", \"7x50:7x55\",\"7x52:7x53\", \"7x53:8x50\", \"7x54:8x50\", \"7x54:8x51\"]\n",
    "    res_contact_tup = [(f.split(\":\")[0], f.split(\":\")[1]) for f in res_contact_list]\n",
    "    #учитываем отсутствие некоторых остатков по маппингу\n",
    "    #PP: hard to read command\n",
    "    res_contact_list_inc_map = [f for f in res_contact_list if f.split(\":\")[0] in inv_map_dict.keys() and f.split(\":\")[1] in inv_map_dict.keys()]\n",
    "    #создаем словарь с картой контактов\n",
    "    contact_dict = {}\n",
    "    for item in list(set([f.split(\":\")[0] for f in res_contact_list_inc_map])):\n",
    "            contacts = [f.split(\":\")[1] for f in res_contact_list_inc_map if f.split(\":\")[0] == item]\n",
    "            contact_dict[item] = contacts\n",
    "    #создаем словарь с картой контактов остатков\n",
    "    contact_dict_transl = {}\n",
    "    #PP: hard to read command\n",
    "    for key, value in contact_dict.items():\n",
    "            contact_dict_transl[inv_map_dict[key][1:]+\"_\"+inv_d[inv_map_dict[key][0]]] = [inv_map_dict[f][1:]+\"_\"+inv_d[inv_map_dict[f][0]] for f in value]\n",
    "    #создаем список остатков, которые образуют контакты\n",
    "    residues_inv_in_contacts_list = []\n",
    "    for item in res_contact_list_inc_map:\n",
    "            residues_inv_in_contacts_list.append(inv_map_dict[item.split(\":\")[0]][1:]+\"_\"+inv_d[inv_map_dict[item.split(\":\")[0]][0]])\n",
    "            residues_inv_in_contacts_list.append(inv_map_dict[item.split(\":\")[1]][1:]+\"_\"+inv_d[inv_map_dict[item.split(\":\")[1]][0]])\n",
    "    residues_inv_in_contacts_list = list(set(residues_inv_in_contacts_list))\n",
    "\n",
    "    if [f for f in res_contact_list if f not in res_contact_list_inc_map] != []:\n",
    "        warnings.warn(\"Those features ({}) won't be computed because of missing mapping labels in mapping file. Check and recalculate mapping file or add missing mapping labels manually.\")\n",
    "    \n",
    "    fi = open(PDBFile, 'r')\n",
    "    all_lines = fi.readlines()\n",
    "    fi.close()\n",
    "    atom_lines = [l for l in all_lines if l[0:6] == 'ATOM  ']\n",
    "    dict_coord = {} # dict to store coordinates. dict_coord[res][atom] = (x,y,z,occ)\n",
    "    atomnum_2_name = {} # map atom number to atom name, in order to find N, CA, C, O\n",
    "    contact_score = {} # dict to store final results. contact_score[ires][jres] = contact_score.\n",
    "    for line in atom_lines:\n",
    "        # retrive info from each atom line\n",
    "        atom_num = int(line[6:11].strip())\n",
    "        atom_name = line[12:16].replace(' ', '_')\n",
    "        res_name = line[17:20]\n",
    "        res_num = int(line[22:26].strip())\n",
    "        #chain_id = line[21:22]\n",
    "        x = float(line[30:38].strip())\n",
    "        y = float(line[38:46].strip())\n",
    "        z = float(line[46:54].strip())\n",
    "        occ = float(line[54:60].strip())\n",
    "        res = str(res_num) + '_' + res_name\n",
    "        atomnum_2_name[atom_num] = atom_name\n",
    "        if res_num <= 0:\n",
    "            continue\n",
    "        if res not in dict_coord:\n",
    "            dict_coord[res] = {}\n",
    "        dict_coord[res][atom_num] = (x, y, z, occ)\n",
    "    mean_coords_res = {}\n",
    "    for ires in residues_inv_in_contacts_list:\n",
    "        atom_in_ires = dict_coord[ires].keys()\n",
    "        #print(atom_in_ires)\n",
    "        coords = []\n",
    "        for iatom in atom_in_ires:\n",
    "            iatom_name = atomnum_2_name[iatom]\n",
    "            #PP: not clear\n",
    "            if (ires.split('_')[1]) == \"GLY\":\n",
    "                if iatom_name in ['_N__', '_C__', '_O__']:\n",
    "                        continue\n",
    "                (ix, iy, iz, iocc) = dict_coord[ires][iatom]\n",
    "                coords.append([ix, iy, iz])\n",
    "            else:\n",
    "                if iatom_name in ['_N__', '_CA_', '_C__', '_O__']:\n",
    "                        continue\n",
    "            #print(iatom)\n",
    "                (ix, iy, iz, iocc) = dict_coord[ires][iatom]\n",
    "                coords.append([ix, iy, iz])\n",
    "        #print(x_coords, Average(x_coords), y_coords, Average(y_coords), z_coords, Average(z_coords))\n",
    "        coords = np.array(coords)\n",
    "        #print(coords)\n",
    "        mean_coords = np.mean(coords, axis=0)\n",
    "        mean_coords_res[ires] = mean_coords\n",
    "    inter_df = []\n",
    "    for ires in contact_dict_transl.keys():\n",
    "        for kres in contact_dict_transl[ires]:\n",
    "            ires_num = int(ires.split('_')[0].strip())\n",
    "            kres_num = int(kres.split('_')[0].strip())\n",
    "            #PP: not clear; mc\n",
    "            if abs(ires_num - kres_num) < 5:\n",
    "                icor = mean_coords_res[ires]\n",
    "                kcor = mean_coords_res[kres]\n",
    "                distance = np.linalg.norm(icor - kcor)\n",
    "                inter_df.append({\"Res_1\":d[(ires.split('_')[1])]+(ires.split('_')[0]), \"Res_2\":d[(kres.split('_')[1])]+(kres.split('_')[0]), \"{:}_{:}\".format(pdb_id, PDBFile.split(\"/\")[-1].rstrip(\".pdb\")): distance})\n",
    "            elif abs(ires_num - kres_num) > 4:\n",
    "                icor = mean_coords_res[ires]\n",
    "                kcor = mean_coords_res[kres]\n",
    "                distance = np.linalg.norm(icor - kcor)\n",
    "                inter_df.append({\"Res_1\":d[(ires.split('_')[1])]+(ires.split('_')[0]), \"Res_2\":d[(kres.split('_')[1])]+(kres.split('_')[0]), \"{:}_{:}\".format(pdb_id, PDBFile.split(\"/\")[-1].rstrip(\".pdb\")): distance})\n",
    "\n",
    "    #PP: do you need all these dfs?; go though the code and try to optimize            \n",
    "    inter_df = pd.DataFrame(inter_df).drop_duplicates()\n",
    "    inter_df1 = inter_df.replace(({\"Res_1\": map_dict}))\n",
    "    inter_df2 = inter_df1.replace(({\"Res_2\": map_dict}))\n",
    "    \n",
    "    contacts_df = inter_df2[inter_df2[[\"Res_1\", \"Res_2\"]].apply(tuple, axis=1).isin(res_contact_tup)]\n",
    "    contacts_df[\"Interacting_residues\"] = [\":\".join([k,f]) for k,f in zip(contacts_df[\"Res_1\"].tolist(), contacts_df[\"Res_2\"].tolist())]\n",
    "    contacts_df = contacts_df[[\"Interacting_residues\", \"{:}_{:}\".format(pdb_id, PDBFile.split(\"/\")[-1].rstrip(\".pdb\"))]].T\n",
    "    contacts_df = contacts_df.rename(columns=contacts_df.iloc[0]).drop(contacts_df.index[0])\n",
    "    contacts_df = contacts_df[res_contact_list_inc_map]\n",
    "    \n",
    "    one_mod_df = pd.read_csv(one_mod_df)\n",
    "    bi_mod_df = pd.read_csv(bi_mod_df)\n",
    "    \n",
    "    for cf, ft in zip(contacts_df.T[contacts_df.index[0]].values, contacts_df.columns):\n",
    "        #PP: mc\n",
    "        one_mod_feat = [\"2x42:4x45\", \"2x43:7x53\", \"2x45:4x50\", \"3x50:6x37\", \"3x51:5x57\", \"6x44:6x48\", \"6x44:7x45\", \"6x48:7x45\", \"7x45:7x49\"]\n",
    "        if ft in one_mod_feat:\n",
    "            appr_mean = one_mod_df[one_mod_df[\"Feature\"] == ft][\"approx_mean\"].values[0]\n",
    "            appr_std = one_mod_df[one_mod_df[\"Feature\"] == ft][\"approx_std\"].values[0]\n",
    "            l = 1\n",
    "            while ((cf > appr_mean - appr_std*l) and (cf < appr_mean + appr_std*l)) == False:\n",
    "                l+=1\n",
    "            if l < 5 and l>3:\n",
    "                warnings.warn(\"Smth wrong with feature {}. Check the distance between common activation pathway residues visually!\".format(ft))\n",
    "            elif l>= 5:\n",
    "                warnings.warn('Ambiguous {} feature value {}. Try to check structure mapping and structure model!'.format(ft, cf))\n",
    "\n",
    "        else:\n",
    "            appr_mean_1 = bi_mod_df[bi_mod_df[\"Feature\"] == ft][\"approx_mean_1\"].values[0]\n",
    "            appr_std_1 = bi_mod_df[bi_mod_df[\"Feature\"] == ft][\"approx_std_1\"].values[0]\n",
    "            appr_mean_2 = bi_mod_df[bi_mod_df[\"Feature\"] == ft][\"approx_mean_2\"].values[0]\n",
    "            appr_std_2 = bi_mod_df[bi_mod_df[\"Feature\"] == ft][\"approx_std_2\"].values[0]\n",
    "            l = 1\n",
    "            while (((cf > appr_mean_1 - appr_std_1*l) and (cf < appr_mean_1 + appr_std_1*l)) or ((cf > appr_mean_2 - appr_std_2*l) and (cf < appr_mean_2 + appr_std_2*l))) == False:\n",
    "                 l+=1\n",
    "            if l < 5 and l>3:\n",
    "                warnings.warn(\"Smth wrong with feature {}. Check the distance between common activation pathway residues visually!\".format(ft))\n",
    "            elif l>= 5:\n",
    "                warnings.warn('Ambiguous {} feature value {}. Try to check structure mapping and structure model!'.format(ft, cf))\n",
    "\n",
    "    return contacts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calculates array of model results for trajectory\n",
    "def main_trajectory(pdb_id, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, frames_dir_path, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict):\n",
    "    rf = joblib.load(model)\n",
    "    traj_dfs = []\n",
    "    i=0\n",
    "    #for filenames like frame_NNN\n",
    "    for PDBFile in tqdm(sorted([frames_dir_path+f for f in os.listdir(frames_dir_path)], key=lambda x: int(x.split(\"/\")[-1].rstrip(\".pdb\").split(\"_\")[1]))):\n",
    "        if i == 0:\n",
    "            map_df = GPCRdb_mapping_for_sequence(pdb_id, PDBFile, path_to_gpcrdb_files, dir_path, out_path, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict)\n",
    "            res_array = calc_dist_feature_modif_no_c_id(PDBFile, map_df, pdb_id, one_mod_df, bi_mod_df).to_numpy()\n",
    "            res = rf.predict_proba(res_array)\n",
    "            res_arr = res\n",
    "            i+=1\n",
    "        else:\n",
    "            res_array = calc_dist_feature_modif_no_c_id(PDBFile, map_df, pdb_id, one_mod_df, bi_mod_df).to_numpy()\n",
    "            res = rf.predict_proba(res_array)\n",
    "            res_arr = np.append(res_arr, res, axis = 0)\n",
    "            \n",
    "    return res_arr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#prepares dataframe for trajectory results visualization\n",
    "def trajectory_df_for_viz(pdb_id, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, frames_dir_path, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict):\n",
    "    res = main_trajectory(pdb_id, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, frames_dir_path, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict)\n",
    "    df_res = pd.DataFrame(res, columns = ['Inactive_prob','Active probability'])\n",
    "    #df_res[\"label\"] = pdb_id+PDBFile.split(\"/\")[-1].rstrip(\".pdb\")\n",
    "    df_res[\"frame\"] = [f for f in range(0, len(df_res['Active probability']))]\n",
    "    return df_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#resuts visualization in case of trajectory\n",
    "def traj_viz(pdb_id, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, frames_dir_path, lim, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict):\n",
    "    import matplotlib.pyplot as plt\n",
    "    import joblib\n",
    "    import seaborn as sns\n",
    "    plt.figure(figsize = (30, 10))\n",
    "    df_t = trajectory_df_for_viz(pdb_id, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, frames_dir_path, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict)\n",
    "    ax1 = sns.scatterplot(data=df_t,x=\"frame\", y=\"Active probability\", alpha = 0.2, palette = \"flare\").set(ylim=(0, 1), xlim=(0, lim))\n",
    "    smoothed_dfs = []\n",
    "    list_sele = df_t[\"Active probability\"].to_list()\n",
    "    FIVE_split = np.array_split(list_sele, len(list_sele)//5)\n",
    "    proc_data = [f.mean() for f in FIVE_split]\n",
    "    df_pl = pd.DataFrame(proc_data, columns = ['Proc_data'])\n",
    "    df_pl[\"frame\"] = [f*5 for f in range(0, len(list_sele)//5)]\n",
    "    smoothed_dfs.append(df_pl)\n",
    "    df_s = pd.concat(smoothed_dfs)\n",
    "    ax1 = sns.lineplot(data=df_s, x=\"frame\", y=\"Proc_data\", palette = \"flare\").set(ylim=(0, 1), xlim=(0, lim))\n",
    "    plt.axhline(y=0.375, c=\"red\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def main(pdb_id, PDBFile, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict):\n",
    "    map_df = GPCRdb_mapping_for_sequence(pdb_id, PDBFile, path_to_gpcrdb_files, dir_path, out_path, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict)\n",
    "    res_array = calc_dist_feature_modif_no_c_id(PDBFile, map_df, pdb_id, one_mod_df, bi_mod_df).to_numpy()\n",
    "    rf = joblib.load(model)\n",
    "    res = rf.predict_proba(res_array)\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'N' for Atom (name=N) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=H1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=H2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=H3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CA) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HA) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CB) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HB1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HB2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'S' for Atom (name=SD) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HE3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=C) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=O) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CD) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HN) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HB) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=1HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=2HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=3HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=1HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=2HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HD3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HA1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HA2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=OG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=3HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CZ) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=OH) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HH) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=OG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=OE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=OE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=1HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=2HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=3HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=1HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=2HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=3HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HB3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=OD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'N' for Atom (name=ND2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'S' for Atom (name=SG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'N' for Atom (name=NE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CE3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CZ3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HZ3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CZ2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HZ2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'C' for Atom (name=CH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'N' for Atom (name=NE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=1HE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=2HE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HZ) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=OD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'N' for Atom (name=ND1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'N' for Atom (name=NE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'N' for Atom (name=NH1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=1HH1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=2HH1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'N' for Atom (name=NH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=1HH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=2HH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'N' for Atom (name=NZ) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'H' for Atom (name=HZ1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=OT1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/Atom.py:232: PDBConstructionWarning: Used element 'O' for Atom (name=OT2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/usr/local/lib/python3.9/site-packages/Bio/PDB/PDBParser.py:395: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 5052\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.9/site-packages/Bio/SeqIO/PdbIO.py:303: BiopythonParserWarning: 'HEADER' line not found; can't determine PDB ID.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nadd = 1\n",
      "rescale = 1\n",
      "dndpre (aa) Version 7.453\n",
      "alg=X, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "rescale = 1\n",
      "All-to-all alignment.\n",
      "  288 / 289\n",
      "\n",
      "##### writing hat3\n",
      "pairlocalalign (aa) Version 7.453\n",
      "alg=Y, model=BLOSUM62, 2.00, -0.10, +0.10, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "nadd = 1\n",
      "nthread = 0\n",
      "blosum 62 / kimura 200\n",
      "sueff_global = 0.100000\n",
      "norg = 289\n",
      "njobc = 290\n",
      "Loading 'hat3' ... \n",
      "done.\n",
      "rescale = 1\n",
      "Loading 'hat2n' (aligned sequences - new sequences) ... done.\n",
      "Loading 'hat2i' (aligned sequences) ... done.\n",
      "c / 1                    \n",
      "\n",
      "Combining ..\n",
      "   done.                      \n",
      "\n",
      "   done.                      \n",
      "\n",
      "addsingle (aa) Version 7.453\n",
      "alg=A, model=BLOSUM62, 1.53, -0.00, -0.00, noshift, amax=0.0\n",
      "0 thread(s)\n",
      "\n",
      "\n",
      "To keep the alignment length, 18 letters were DELETED.\n",
      "The deleted letters are shown in the (filename).map file.\n",
      "\n",
      "Strategy:\n",
      " Multi-INS-full (Not tested.)\n",
      " ?\n",
      "\n",
      "If unsure which option to use, try 'mafft --auto input > output'.\n",
      "For more information, see 'mafft --help', 'mafft --man' and the mafft page.\n",
      "\n",
      "The default gap scoring scheme has been changed in version 7.110 (2013 Oct).\n",
      "It tends to insert more gaps into gap-rich regions than previous versions.\n",
      "To disable this change, add the --leavegappyregion option.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "./test/identifier1_mapping.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.9/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator DecisionTreeClassifier from version 0.24.1 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n",
      "/usr/local/lib/python3.9/site-packages/sklearn/base.py:329: UserWarning: Trying to unpickle estimator RandomForestClassifier from version 0.24.1 when using version 1.0.2. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:\n",
      "https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[0.32, 0.68]])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main(pdb_id, PDBFile, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "traj_viz(pdb_id, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, frames_dir_path, 2500, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "single pdbs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_id = \"2YDV\"#file identifier\n",
    "PDBFile = os.path.join(working_dir,\"Files/examples/Single_pdbs/2YDV_frame_22_act.pdb\")#path to single pdb file\n",
    "path_to_gpcrdb_files = os.path.join(working_dir,\"Files/resources/Gpcrdb_files/\")\n",
    "one_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/onemodal_features_dist_stats.csv\")\n",
    "bi_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/bimodal_features_dist_stats.csv\")\n",
    "model = os.path.join(working_dir,\"Files/resources/RF_classifier_good_model_19.joblib\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_id = \"4EIY\"#file identifier\n",
    "PDBFile = os.path.join(working_dir,\"Files/examples/Single_pdbs/4EIY_frame_92_inact.pdb\")#path to single pdb file\n",
    "path_to_gpcrdb_files = os.path.join(working_dir,\"Files/resources/Gpcrdb_files/\")\n",
    "one_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/onemodal_features_dist_stats.csv\")\n",
    "bi_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/bimodal_features_dist_stats.csv\")\n",
    "model = os.path.join(working_dir,\"Files/resources/RF_classifier_good_model_19.joblib\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_id = \"6D9H\"#file identifier\n",
    "PDBFile = os.path.join(working_dir,\"Files/examples/Single_pdbs/6D9H_frame_6_act.pdb\")#path to single pdb file\n",
    "path_to_gpcrdb_files = os.path.join(working_dir,\"Files/resources/Gpcrdb_files/\")\n",
    "one_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/onemodal_features_dist_stats.csv\")\n",
    "bi_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/bimodal_features_dist_stats.csv\")\n",
    "model = os.path.join(working_dir,\"Files/resources/RF_classifier_good_model_19.joblib\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Пример с подвохом, как раз тут вылетает Х\n",
    "pdb_id = \"4Z34\"#file identifier\n",
    "PDBFile = os.path.join(working_dir,\"Files/examples/Single_pdbs/4Z34_frame_43_inact.pdb\")#path to single pdb file\n",
    "path_to_gpcrdb_files = os.path.join(working_dir,\"Files/resources/Gpcrdb_files/\")\n",
    "one_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/onemodal_features_dist_stats.csv\")\n",
    "bi_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/bimodal_features_dist_stats.csv\")\n",
    "model = os.path.join(working_dir,\"Files/resources/RF_classifier_good_model_19.joblib\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=N) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=H1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=H2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=H3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CA) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HA) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CB) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HB1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HB2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'S' for Atom (name=SD) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HE3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=C) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=O) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CD) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HN) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HB) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=3HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HD3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HB3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CZ) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HZ) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OH) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HH) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HA1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HA2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=3HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=3HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=3HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=ND2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CE3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CZ3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HZ3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CZ2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HZ2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NZ) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HZ1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NH1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HH1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HH1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'S' for Atom (name=SG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=ND1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OT1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OT2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/PDBParser.py:395: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 5161\n",
      "  warnings.warn(\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/SeqIO/PdbIO.py:303: BiopythonParserWarning: 'HEADER' line not found; can't determine PDB ID.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/madara/ilya/GPCRapa/test/6D9H_mapping.csv\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([[0., 1.]])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "main(pdb_id, PDBFile, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "trajectories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#пример с Х\n",
    "pdb_id = \"3REY\"#file identifier\n",
    "frames_dir_path = os.path.join(working_dir,\"Files/examples/Trajectories/35_10386_apoform_3REY/\")#path to trajectory pdb files\n",
    "path_to_gpcrdb_files = os.path.join(working_dir,\"Files/resources/Gpcrdb_files/\")\n",
    "one_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/onemodal_features_dist_stats.csv\")\n",
    "bi_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/bimodal_features_dist_stats.csv\")\n",
    "model = os.path.join(working_dir,\"Files/resources/RF_classifier_good_model_19.joblib\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdb_id = \"6DDE\"#file identifier\n",
    "frames_dir_path = os.path.join(working_dir,\"Files/examples/Trajectories/ClassA_oprm_mouse_6DDE_refined_Active_2020-10-08_GPCRdb(short)/\")#path to trajectory pdb files\n",
    "path_to_gpcrdb_files = os.path.join(working_dir,\"Files/resources/Gpcrdb_files/\")\n",
    "one_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/onemodal_features_dist_stats.csv\")\n",
    "bi_mod_df = os.path.join(working_dir,\"Files/resources/mod_data/bimodal_features_dist_stats.csv\")\n",
    "model = os.path.join(working_dir,\"Files/resources/RF_classifier_good_model_19.joblib\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "c5a01567ded44e439eb79753c8b54a95",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(HTML(value=''), FloatProgress(value=0.0, max=201.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=N) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=H1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=H2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=H3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CA) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HA) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CB) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HB) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=3HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=C) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=O) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HN) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HA1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HA2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HB1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HB2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CD) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HG2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'S' for Atom (name=SD) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HE3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=3HG1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HB3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HD3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=3HD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=3HD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CZ) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OH) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HH) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'S' for Atom (name=SG) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HZ) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OD1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=ND2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HE) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NH1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HH1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HH1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NZ) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HZ1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HZ2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HZ3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OD2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=1HE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=2HE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=NE1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CE3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CZ3) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CZ2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'C' for Atom (name=CH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'H' for Atom (name=HH2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'N' for Atom (name=ND1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OE2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OT1) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/Atom.py:218: PDBConstructionWarning: Used element 'O' for Atom (name=OT2) with given element ''\n",
      "  warnings.warn(msg, PDBConstructionWarning)\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/PDB/PDBParser.py:395: PDBConstructionWarning: Ignoring unrecognized record 'TER' at line 4856\n",
      "  warnings.warn(\n",
      "/home/madara/anaconda3/envs/modelling/lib/python3.8/site-packages/Bio/SeqIO/PdbIO.py:303: BiopythonParserWarning: 'HEADER' line not found; can't determine PDB ID.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/madara/ilya/GPCRapa/test/6DDE_mapping.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-5-f3773f7fb03a>:123: UserWarning: Smth wrong with feature 3x51:5x57. Check the distance between common activation pathway residues visually!\n",
      "  warnings.warn(\"Smth wrong with feature {}. Check the distance between common activation pathway residues visually!\".format(ft))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABsoAAAJRCAYAAADs7uJCAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8li6FKAAAgAElEQVR4nOzda3BcaX7f99+59TndpxuNA4AACDZvQ3KGM5yZ3eGMdlcjab3SelertZS1bCmRIluOEltlyUpcSakSRYkTpZRLpRK7yrKkVUmy5Ti6xFpFUqTVdWWtZO1tZnbul+XskMMbSIIkgEaj0ddzTp+8aPRBNy4kQBIDgvh+qljo85zn8n+ew3f/ep7HSJJEAAAAAAAAAAAAwF5j7nQAAAAAAAAAAAAAwE4gUQYAAAAAAAAAAIA9iUQZAAAAAAAAAAAA9iQSZQAAAAAAAAAAANiTSJQBAAAAAAAAAABgTyJRBgAAAAAAAAAAgD1p2xJlhmH8K8MwbhiG8cYG7w3DMH7GMIyzhmG8ZhjG6e2KBQAAAAAAAAAAAFhtO3eU/WtJn7jF+++QdGL53w9L+vQ2xgIAAAAAAAAAAAAM2LZEWZIk/17S/C2qfErSv0m6viJp2DCM/dsVDwAAAAAAAAAAANBvJ+8oOyDpct/z9HIZAAAAAAAAAAAAsO3sHRzbWKcsWbeiYfywusczyvf9p0+ePLmdcQ2IOonCqLOm3LFN2eZ6U7g3bQEAAAAAAAAAADDoxRdfnE2SZN+96m8nE2XTkg72PZckXV2vYpIkvyjpFyXpmWeeSb761a9uf3TLqs1QF2bra8qPjOVU8JxtawsAAAAAAAAAAIBBhmFcvJf97eTRi78n6QeNrg9JqiRJcm0H41lXwXMU+INJrcB3NpXoupu2AAAAAAAAAAAA2F7btqPMMIzfkPQRSWOGYUxL+h8lOZKUJMkvSPpDSZ+UdFZSXdIPbVcst1NthmqFsVzHWjeJVQpyKmZvXWcjd9N2q3ECAAAAAAAAAABg87YtUZYkyfff5n0i6R9t1/ibNV2uq1wL0+fAd1QKcmvqFbw73wl2N217NhsnAAAAAAAAAAAANmcnj17ccdVmOJB8kqRyLVS1GW7QYmfsljgBAAAAAAAAAAB2kz2dKGuF8ZbKd8puiRMAAAAAAAAAAGA32dOJMtextlS+U3ZLnAAAAAAAAAAAALvJtt1RthOqzVCtMJbrWLe8E6xXL+p01A5jzdXbGso6ymdsuRkz3anV62N1v5sd514peI4C31lzR9l7MTYAAAAAAAAAAMCD6oFJlE2X62sSSaUgt269hVqo+VpLl8sNteJYw15GYRwrybtK5Ohau5X2ISnt15DUjjtyLPO249xrpSCnYva9TdABAAAAAAAAAAA8yB6IoxerzXAgSSZ1k1vVZrhuvVYUa3aprbmltpYasZpRrFaY6PzNmmrtKK1/vdLUdLmRPreiWBfn6gN11htnuxQ8R2MFjyQZAAAAAAAAAADAPfBAJMp6RyXerrz33I4ShXGSlkdxR2GcdP9GK23aUbLmWdJA2a3GBwAAAAAAAAAAwP3rgTh60XWsTZX3njO2Iccy0nLbMuVYhuLElGOvtMnYhozO4LOkgTq3Gh8AAAAAAAAAAAD3rwciUVbwHAW+s+aOstVHFPbqLdSksXxGjTBWI4qVdWwNZW0dyfkD949NFD1JK3eUubalw6O5NXeUcRQiAAAAAAAAAADA7vNAJMokqRTkVMyGaoWxXMdKk1fV5mDZQr2t587PqdGOZZnS/uGcHt1fkOdYGvEzkrSmj9X9ru7zdnr1o05HtmmmO9BuF+tWx1mvDwAAAAAAAAAAAKzvgUmUSd0dY/3JoelyPd0NZkgK445+7Dde1pWFRlrHNKR//n1P6cior8VGpMB3VApyt+x39fOtTJfrWqiFmqu1NF8LNZS1ZZuGwk6iiUJ3x1rgd/vqj7Udd9bsXFsd13pjrd5Vd7s2AAAAAAAAAAAAe5V5+yq7U7UZDiSNWlGsr11b1JWFhk4fHNYPfuiwvu3kPnUS6e2ZRdXakaRusqraDDfq9o5iaEWx5pdjqdRDXVloqNqI0jGvV5qaLq8k71pRrItz9fT9ZuJaPd97PRcAAAAAAAAAAIAHzQObKGuF8cBzO0p0Ya4uSToy5mss72r/UFaSNFttKYziDdtKUpIkdxxDO1ppG8aJorjT/R2tvO8fv1e/v2yjuG737lZtAAAAAAAAAAAA9rIHNlHWuwesJ2Mburp85GLvyMOxfPfIw4VGKMe2Nmz7h69f0+mf/pyWWpG2otdPxjbSMscyZC8fqdgbM2MbA+P36veXrRfXZt7dqg0AAAAAAAAAAMBetusSZVEn2dRxggXPSe/+kiTXtjRfb2tf3lVpJCvTkIJ8RrmMpVo7VsY0tFBvqx2vJMOqzVCX52r6jecvqVwP9fXrVVWboWarzS3F4NqWRpZjGc45Gs+7MoyVXWYTRU+lINu9Ry2K1WjHGvJMNduRFuot1drdu9NudS/a6vlKum0bAAAAAAAAAACAvcze6QC2Kow6ujBbV+A7KgW5W9YtBTkVs6FaYSzXsXRpvq7Th4f14RNjulSuq9aKtb/oaX6prTPXFyUZyjq2FuqRRnIZLTZD3ay29cKFeUnS5792Q465klvcSgyTRVdRp6O5pZaWWh0FfkbtKJGXMdM+6q1FXZhraKkVKUmkRhRrPO9qf3Fzn2n1fEmSAQAAAAAAAAAAbGzX7SjrKdfCTe/qGit4CuNE0+WGTh8KNOxnNORltL+Y1eSQp+vVpmarK31V6qHOzCzqykJDF+ZqaobdO8Veu7KgWntlx9lWY/BdR53EVC5jK2Nbynu2mu2Oqs1w+V+kMJYylqWb1baWGrEWm5HCTrLlsUiSAQAAAAAAAAAA3NquTZRJUiuMN133tekFSdKTpeGBdmN5V/O1tjqdRFHcTYiFcaJmGCmKOzozU5VpSEOeoxuLLYXR4JhbiWGjuq0wViuM1Y6SdPyeKO6kY25lLAAAAAAAAAAAANzark6UuY616bqvT1dkGNLjB4YG2k0Ne+okUrUVyba6y+FYhjzHlm2ZOjOzqCOjvg6OZDVXa8uxB8fcKIYkSfQv/t07euvq4m3ruo4l17GUsY10/B7bMtMxtzJfAAAAAAAAAAAA3NquTJQZkryMqVYYa6ZS12y1mR5f2Pu92qvTFR0Z9dNdWYHfPZrw4PL9YK7dvZ/MMKR81lIp8GRbhqbLDZ2cLOjQSE5ztbZa7UiVRltLzUhexlTBczRTqeuN6bLOXFtIx379SkX/9HNf19/7led1ca4mqXssYm/cnsB3VPB6/2w5ltSOY43mM8p5lsaHPOUz9rrz3Ypbrc3duNt+76b9ds0JAAAAAAAAAADsDfZOB7BVjm0qYxtqtTt6s7yo+Vqooawt2zQUdhJNFDxJ3QRUaTkJJkkvXy7riQNFXau00vdHxnKSusct7hvy9MzhYV1fbClKEuUztr56oXtc49986oDOzCzqj96Y0XS5IRmGhn1HQ56tP3/ruq4tNjW31JYkTQVVHRvL63deuSJJmltq6Ud+9SV9+u+c1uFRX6Ugp2I2VCuM5TqWCp6jP35jRq9NL6jRjhV3OspmbD17bESPTBZkm6bma+2B+RaytiYK3po5bmS6XFe5tpJM2my77e73btpv15wAAAAAAAAAAMDesesSZaYhtaJE7SjW/HKipFIP1YxiZR1beTeSn7FVroUqZkMVPEfnblQ1t9TWifFC2k/3vaPHpoblWIZuVFsa9jNaaERyJSXq7kIrZh09fqCghXp3rPNzdZ2aKqodJrpZa+rM9UXVWrHc5eMRbyw2VW+GeulSWcM5Rx9/dFK/+eJl/cy/e0c/9R+c6ts91t1Zdnm+rn/4qy+umecvf+G8Pnh0RJ98YlJPHhhWGHfS+VYbkfJuJEnpHDdSbYYDCaWVud+63e3cbb9303675gQAAAAAAAAAAPaWXZco6yTdv+0oScvCOFEUdyRHCqNYynSn1QpjFTxHL1/q7gw7MZ4f6Kv3vhTkdGm+nh7L2B0n0cuXF/TUwWHVmrEODGclSTeqLZ1arrNYDxXGUtzpSLKWY+ned/buzZoe2z+kDz00ogtzNf2/L13Rh0+M6VNPlQZi+MyL0zIkffoHntZoPiNJujRf15fOzem583P673/3Tf3kJx/VE1PFgXa9efbmsJH+Oa039zt1t/3eTfvtmhMAAAAAAAAAANhbdt0dZabR/ZuxjbTMsQzZVncqzvLOLklyne7vr99YkmlIR8f8gb567w+O5HR5vp4+S9K7N2uqNEKdPhyokLU1lO0etXij2krrDOUcOZZkmSvL6FhSrRWp3o51YqIgwzD0t0+XdGgkp//5D86o2ZfkiTuJfuurl/XssVEdCLLyHEueY+nhiYL+k2eP6Pd/7Ju0L+/qc2/NDMy3f579Ma9no/e3a3c7d9vv3bTfrjkBAAAAAAAAAIC9ZRcmygwFviPXtjTid3cPFXOODgxnVcja8pd3kwX+yvGGZ2aqOjrmy+tLpPS/PzSS1cW5ugqeo2C5zxcvlSVJ3/rIPk0Wc+kdWPO1bqKskLW1z/d0cmJIU8u7zSRpfMhTc3m324l93R1so4WM/utPPKKbSy396lcupnW/eHZWVytNff8HD6Xj9scX+K6+55mSXrxY1lIzSufbm2f/HDbSP6f15n6n7rbfu2m/XXMCAAAAAAAAAAB7y647elGSillHSSeRYxuaLHryHEsjfvfYwlYYK+p0ZJumZip1WYahVy8v6CMP71PeNRXFiQpZW5PFXNrf4RFflUaoSj1UKcipmA312vSCHt1f0BOlYVWboTzb1OHRrL787rzef3BIBc+R61g6NJbTzEJDlWaovGvpQODrs69d04FhT990YlSS0vi++fiYfv7zZ/Utx0flZWz92nMXVcw6OjqS0/VKQ7YlDec8ubYh2zRVbYb65OOT+vRfnNNfvnNTf+/ZI3pk0lAYJaqHsWqtSBfnahrxM2uSRDOVuqqNSIWsnc6pFcZyHSutW22GWqi11ZHSPqrNtfV6+t/VWqEarUh511TBcwbW3DbNtH1vjHKjLUkaL7jp2ve+Y//46421XgJsozkBAAAAAAAAAABs1q5LlLXjji7O1jVXa2m+FqqQtTVR8GSZhkpBTpVGqIVaqLlaS+V6qOuLTVUaoRzL0CuXF7v1I09Rp65S0E3YHBzp/r1crquYK6rajPTadEU//vGHNV2up/3ZpqWFeqiZSluObakZdVSuhZIk17aV9xzN19p64fy8vvn4mGYqTYWdRBMFT9VGpO98Yr++cHZWP/Pn5/TssVH92ddu6INHA/3pW9c1t9RW1rV0MMhqKshpsuBpvtbSXK2th8Z8/f6r1/TssVGN+q6qzVCXyw3NLbWV8ywdGfV1crKQzueli2VdnKuna3Z4tNU9QrIvmTRdruvcjSXNL8c/lLUV5DJyrJVNhr1ddL365VooU9K7s0u6vtTSaM6VYUgF19bEkDfwTSYLnsK4o6VmqK/fXNL5GzW5GVP7h7N6+mCgfUNuunaSZJlGGl9vrPXi6Ffw2EUGAAAAAAAAAADu3K5LlMVxolYUpwmeaiNS3o0kSbbZTbC0l98nSaI/P3NDhqQgl1Ej7Nbr1S9mQxU8R4eWE2UX5+p6/EBRf/DaNUnStz4yPtDfeMGVJJ29WVXGNhR2kvSoR0m6XmnqzPWqau1Yj00NpTHm3UiOaShjm3pkoqDPv31DcSdR3Ek0ksvoWqUpxzIVx4neuLKoTieRZ5uaXWprutzQ6UPD+q2XruiVSws6EHSPeZxb6u7Qqjdj3VhsKu/aKmYd1VrhQJKsN6+p4ZWdXNVmqOuVZhqfJFWWk4oP7cuncyrXQhWzYfq71/b8bLf/nBPJsy29eXVRUkeLzU76TTy7rRuLLbWjWOdv1JRIarY7Ki+1dWZmUZWGp9G8l46/3lir35EUAwAAAAAAAAAA99Kuu6NMktrLd4D1hFEsqZug6X/fijp68+qiju3zu8cDxp2B+q2w+/fQaDeBdGm+mwD67GtX9cSBoiaG3IH+xpefL8/X1Y6StJ/+uF67vCBJeni8MBBfO0oUxom+9ZEx1dux/uTNGY0XXOU9W3GnG1c3edZRI4xVb4UK4+64JycLcm1TX353Xs0wUnM54dcTxR2FUaxWGKdrsFp/eSuM165hnKT99GuFcbpOklRrr/yO4k4aY2XVuL34l1qh+keKOx01w0i11to4V4+1+h0AAAAAAAAAAMC9tCsTZRnbGHh2bEuSVMjaA+/fnV3SUivSB46OSJLs5WMFe/Vdp/s379oa9TO6NF/XxbmaXp2u6Lvetz993+tv1HdlGtL1aksZ20j76Y/rzExVB0dyaVKtN17GNuRYhg6N+nryQFGJpKcOFpWxTFlmNy7LNGSZprKOpZzryLG64+ZcW08dCvT6lYokQ54zuBHQtkw5ttW9qyu7/ibB/nLXsdauoWWk/fRzHStdB0nyMyu/bctMYyyuGneuFuq3Xrosz7HUP5JlmvIcW767Ns7VY61+BwAAAAAAAAAAcC/tukSZZRlybUsjfvcYvqGsLcc0ZBiJfNdR4Dvp++fPz8vPWHrywJDGCt36hawtP2Mr8Afvtzo4ktPl+bo+u3zs4t94ckoFb7A/yzQ0PuTpxmJLE0VPpeVjEHtG8hm9fb2q9x0oyrMtFVxLhtHdT+XZlqYCT0OerY89Nq5nDgf6yMl9emifr9F8Rsby3B4/MKSDo75yjiXftbQ/cGWbht5XKqodd3Sz2taB4axG8xlJUs6zND60EksYJSq4VnrMpCQdGc3Jdx3NVpuaqdTVCmMVPFsjviPDkJIkkesYemifL8c0tNSM1I7igTVqh7FmFhvqJImmhj05tqGsY8s0DZ2aGtLEUC79JoWsra+8O68XLpR1tdLQRNFVp9ORlzEV5DN6dHJIU8NZLdRbqrW7cQbLbVthLHdVEm/1twIAAAAAAAAAALgXdt0dZRnL1OGxnCaLrqJOR3NLLVVbHWUSQxdm6wp8R4fHcvIcQ29eXdR3Pjmp0ogv2zAlJcpnbR0eza1JvBwayenly2XNLrX09OFAB4a7iadSkFMxG6bjPTKR13S5oVLQPa7xxYtX9FdfvynLNFQPO2qGHX300XHJ6KiQdRT4GYVRosV2qELGUaZoajib0aP7Cyp4GWVMQ5XRUIk62l/M6kDg69pCQ9cXmwpyGdmmocVmpA8cDfTbL7n62syifuRbj+nomK96GMtzLI34GVUaoV66WNZ8LZRpSkHOUZDL6MhYTlFHujhb11ytpflaqELW1mTB01SQVT5ja67eVjHrqBXGmmu3NeTaMjrdHVzT5brO3VjSQiNUtRFqsRnq6Liv0vCwZEhHxnKaLOZUba6skW2a+t//6Iwk6eVLFX33+6c0NZzTvoKjhycKijrSQi1ULmOrHSXyMt187YXlu88MqZtUyzrdXXIkyQAAAAAAAAAAwDbYdYkySSp43R1G1Waom9VQfsbUq9MLsk1Dp6aKKmYd/eU7c4o6iT56clLFbCZtmyTr93l4NKffe/WqJOmnvuuxdceTpJP7h/Slc3OK4o7+7Vcv67/7nTdUcG1l7G6y58R4Xh84OqL5WqhcprdhL9al+aaMYUN+xlZW0s2ltvKeZNuWRvPdpNSB5eRbK0o0nHPVjmItNmNJhnzX0cOTBb11bVGSdHDUT+OrNkNdrzQ1XwslSZ2O1OwkktHdmVWuhWpHcfq+2oiUdyO1IkNhkmh/Mat2FOvq8nvHt+RnbF2vNNUIu+2SJEnbl5dCBVlXfsaW7zpr1mim0tC5m0uSpLeuVfW9T1syDEMZ207jkaSMbSljS4v1UHOdtvxM930iqdnuyB0iSQYAAAAAAAAAALbPrkyU9bTCWJIUxh39L3/wNTXCWB9/bEL/7Xc8ot984bIe2z+kI2P+uu1WJ2AOjnSTVIYhffKJ/RuOeWxfXmGc6H/7ozP6l184r287Oa6f/4HT8vru0JqtNgfatKNudi6MYml5F1X/8+r5rG7Xq3tiPK8vnp3V9UpjIP5WGA/U7W9TbURr+uq9S0xLUWcwptVx1lvdpFYYr7yP4k5aZ721/OqFeXUS6elDgV68VNblckOHRnID8ayeZy+Ofuv1DQAAAAAAAAAAcK/sujvK+rnLyanXpytqhLGeORzoz752XX/r01/W29er+tunD9yyXb9Dy4myDx0d1fiQt+GYx/Z1E2//8gvn9e2nJvQLf+fpgSTZev1nlu/ccmxr3ef+dv1tM313dTm2pRPjeUnSOzdqa9plVt3r1WtjGN3jE1e/d+xum9UxrY4z5zqaW2qp3l5JcNmWmdZ558aS2lFnoO/XryzKNKRPPjEpw5DeuFJJ+y1k1+Zm++NYPS8AAAAAAAAAAIDtsqsTZQXPUeA7eu7CvFzb1E98x0n94t99WhNDnoY8W9/7DQcV+IM7kgLfWXeX0vHxvDKWqe95unTLMY+PF5SxTf2NJ/frZ//j0+mRi+vF1ePalg6P5tKjBVc/98fV39a1LY34jgpZW37G1vHlRNnZG0trxuso0UjfmIWsrVKQ1WdeuqIf/8yrsk0zfd/rb6LoqRRk1x1LUvr+5//irH7pr95V4DvKeZbGhzz5GVsX5pb0/b/0nH75C+8OxPPadEXHx/M6Pp7XQ2O+3rxaSeOZLObWfJP+OFavBwAAAAAAAAAAwHbZ1UcvVpuhXMvQVy/M69ljo3pksqCC5+gjj4yr1o415Dka8hwVs6FaYSzX2fjOq7G8q+d+8qMazt06OVPMOvryT3ybHMvQzEJDHUkjfibtt9oMtVBrqyNpX8GRbZrpzqheea9+tbl+XKUgl8Z8eGz53rIwVtRxdHg0p+cvzGu22kzbffncnL7/l76if/o9T+j4hK/hnKcRv3sv2+vTFdXasephqFNTQ4o6HdmmqajTURQnyrm2illHC7W2nOVdZZ5jpe0/88JlzddDSaGqzVDf8fikCp6jjG3qpz/7liTpN1+4rB/5a8e01Iq01Az18qWyvv+Dh3T6cKBPnJrUz/3FOR0Y9lQKcpqp1NVoRbLNRLZpqpC1NVnMLa/tytoVs2vXs3+dAQAAAAAAAAAA7tauTZRNl+sq10KdvbGkG9WW/sGx0TSJYlumitmVnV69nVq3Eywnh26nEcZ648qS5mvd+7sKWVsnJwuSpHM31pY3o47Ky2WSZJlGGtNGca1+V2mEWqiFOjCc1WvTFb0yXdFEwVPgO/q/vnRBkvSV82WZpqUjo6Ys09C5G0s6d7N7TOPn357VwRFfpSCXrt1/9Zuv6OGJgv7+Nx/VYjNM4x7K2hrJZVRptPUrX7qgsXxGhmHo33zlkp49tk+ObelrM1U9f2FeHzg6oufPz+sPXr+mwyO+zswsqhl19NCYr4Ln6HufOaif+4tz+vK5eTXaHV2ar6vWijS31Nb+YU8nJ4cUdeoqBbnuHJfvMKs2IoVxZyCu3nqWgtymvhMAAAAAAAAAAMCt7LqjFxthrGozTBNPXzk/J9OQHt1fVLUZ3qb13as2Q12vNNPkjdRN6lyYrenCbG3d8ulyY6CPci3cUqy9+baiWOMFT5VGqOn5hmrtSJdna/ryu7OSpNmlliTp/GxNb1+raG6prflat+zdm904ZirdJNlSK9I7N5b0+pUFnZ+t6crCSoyVeqgzM4v68rtzulxu6H2lYX3wyIhmKk396Vszml9q6//447c1VfT0S3/3GfkZS7/90hVJ0ltXFyVJh0d9VZuhjoz5OjGe1x+8flUX5+rqdBLNLbUlSdcWmppbaqpcC9O4elpRvCauaiPSdLnxnnxnAAAAAAAAAADw4Nt1ibJ3b9b07s2VO7qee3dOj+4fUjHrqBXG2z5+K4zVjpI15fVWqHprbQKn3goVRmvj2kqsvbrtKNGhke5uqsvzdYVRrLdmqqos78KaXU5AhXGiSiPSYjNSGHdjvbrQUBjFqi7XPbe8hlfKTTXasaK4k44XxomaYaS/emdWWcfSY/uHdGI8r8khT7/14rS+cn5er12p6D//6AkVc44+fmpCXzw7q3o70lvXFjVV9BTkMmncH3tsQi9fWlC9vRJPT621soOsXztKFMbJQFySFEbxe/KdAQAAAAAAAADAg2/XJcok6Z/87puK4o5mFpu6MFfXB4+OSFJ6F9h2ch1LmeW7vPrlXEc5d+0xijnXkWOvjWsrsfbqZmxDB4azMg3p0nxdjm3pzauLMiQdGfXTHWWOZaiYtTW3/Jx3bV2tNOTYlgrZ7mmb71zvJsriJNF8vS3bWvmv4FiGGu2Ovnatqm98aESOZcowDH38sXHdqLb0zz73tkpBVt/zdEmS9L1PH1Qr6ugvv35Tb11d1GNTQwNxf+yxCXUS6Y/fmJGhwUSZ73bj6cXVk7ENOZYxEJckObb1nnxnAAAAAAAAAADw4Nt1ibJSkNVrVyr6zIuX9fz5eUnSB4+OKvBX7vSqNkPNVpuaqdQ1W22q2gzTsrs9tq/gOZooehrxV5JihaytI2O+xguuHEtqhNFAeSnIDvTRH+tmxwx8R65tabLoanLI07XFhjKmoZculXV8PK9S4Gl2qaUkSXR0zNcj+4tqLu9kOzU1pIV6qCTpyDZNubahd25UlVlOQi012ypkLFWbbTXCSMWcozPXqzIM6dljo/I9S45t6MMP79OJ8byaYUc/8teOyVlu//iBIR0Msvr15y6p2op0bF9eXsZM5/i+0rD+w2dK+quzs/o/P/d1LTS6O9/2D3say3vyMqbCKFErjFVrd9fOtS0dHfN1YHhl7YaytkbzGS3U2ro4V7un3xUAAAAAAAAAAOw99u2r3F+KWUff/sFD+rXnLmlyyNNDY76ePT6aJmWmy3Ut1ELN1Vqar4UaytqyTUNhJ9FEwZPUTVSVgtwdx1AKcipmHS3U2upIGvEzqjRCtaNEU8WsitlYOdfUwxNDaVzFbKhWGMt1rC0lyQbHDDVZdPX0kUB/+fZNvTWzqAtzdX3q/VPKOqa+cDbRiQlfT5QCSZJtGjIkfcORQM+dn9cX3plXrd3RZMHTuZtL+oYjw3r+QlmX5hs6ub+oIOeomM1o2M/o86jOdFcAACAASURBVGdu6sMn9unYuC/JUM621OrE+k+/+ai+fG5Op6aKmi7XJUnnbizp6SOBfvflq5KkyaKrZruj6XJdpSAn0zT0X3z0hB6dHNK/+PxZ/coXL+g7n5zUdz81pUYY62q5oflaKMOQshlTedfS4VFfBc9RtRmm69xaVTdJEhWyzj37rgAAAAAAAAAAYG/ZdTvKJOmffOdjOjlZ0MxiU594fHJgJ1m5FqoVxZqvdXcYVeqhriw0VG1E6W6lci28JzvLDo76Ojzqp30m6h4NOJzLKGPZa+qPFbw7SpKt7uP9pWEtNiP96Zs3JEmn9hdVzGYkrdxTJkmXyw2NFVy9vzQsSbpa6a7D1UpD1xdbOjKW16GRnM7N1iQZijqGMo6lF86XVW1F+uuPTmgsn9VY3pNtm5pbinR0zNc/+JaHZJmGrleaujBb03wt1OmDgUxDymUsZSxTtXaUrnPvuzxzZESf/oHT+sTjk/r912b0H/3SV/TFs3Ppt0oSqd7qaKkVD8z54KivET+jajNK63Y6iS7PN3RjsXlPvysAAAAAAAAAANg7dmWizHMs/dwPnNZTh4b1t06X0vJW2E2wtKOVe7DCOFEUd7q/o3hN3Xtho77u5Rj9Hp7IS5K+dG5WE0Oe9hVcjeVdSdK7N2tpvUvzdU0OecplbPmupasLDUnS12cWJUmHR7I6GOR0daGZtgmjWK9NLyhjmXp0/1Ba3lvT/jVsR4nqrW5iyrUtPXN4RO8vDSvuJGm9VhgPrEMuY+tHP3Jc/+t3PyFJ+p8++5Y+8+JlJUnfN4viNWvXCuM131WSorizbd8VAAAAAAAAAAA82HZlokySju3L63d+9Jt0fDyflrmOJUnK2EZa5liG7OW7tBzbWlP3Xtior3s5Rr9TU0XZpqGok+iJqW4yK8hlZBmGZhZXkl5XFxqaLHpyHVNTxayuVbrvLs43ZEg6PlHQ0TFfS60o3Ynl2JZeu1LRk6UhZeyV/x69Ne1fw4xtKOd2d8g5lqFvPj6mDz00Ktsy03quY627Dk8cKOrX//4H9LFHx/Wlc3OaLjfSd47dbXPu5pJevbyQ9rP6u0oaGKtXDwAAAAAAAAAAYDN2baJsPQXPUeA7cm1LI/7y3WA5RweGsypkbfmZ7nGIge/c1RGIG43b716P0W8k7+rE8q6yxw8UJUnDvqP9w56uLifD6u1Is0ttHR3LybUtPbTP17VKQ75r6eJcXYdGcnp4oqATy4nGKwsNFbK24jjR+dmaPvLI+MCcXNvS4dFcuoaSNFH0dGTM14jvyDQNjeYzynmWxoc8+Rk7XYON1ufIWEH/2bccTceXpELWVinIquA5+qnfe1P/5b99pVvuOZooeul3NU1DB0ey6Vi9PrdrzQEAAAAAAAAAwIPHvn2V3aUU5FTMhposuoo6Hdmmme4yaoWxXMfaMJlSbYZqhbGiTkdhlKgjacTP3La+61gqBTnZZl3VRqRC1tZkMXfbWPvbbzXB85FHxlVrxfrup6Ykw9CIn9Hx8bwuznWPXrw83008PX5gWIfHcvrQ0RH9yZvXNeI7end2SR8+sU+lIKdveXhMUnd32PtLRf37d2YlSe87OCzPNrWv4KRrWPAcVZuhFmptdSTZpmSbph6ZzKfr5dqGwihRudHWfK0l25Qmi7n0u6ye7weOjMrPWGpHsU4fHk7Xu9NJ9MrlBTXasa5XGkqUyLMHxxrxu/ey9eIpZkmSAQAAAAAAAACAzXvgEmWS0l1M65VvZLpc10It1HytpcvlhhpRrNGcq0LW1snJgkpBbk39cq17XKEhqR135Cwf8dishoo69TVtNmovdXdD3ar+aj/+8Uf0jz96Ql7fUYOHR3J68UJZSZLo0nw9LSt4jr7h6Kgk6dXpRc0utfVkqbsT7eCIr8khT9PlpsYKnl6+tKBsxlI+Y+tapZXGNlbwJEmVRqhKI9JcraX5WqhC1tZEwVPgOzoc5DRdruuVyws6P9sdP+dZOn0w0OnDwbrfxTQNnTpQ1IX5ug6P+mn5xfm6qs1IkvTFs7MyDHPNWL11XGh06y02oi2vIwAAAAAAAAAA2LseqKMX71S1GapcC9WKYs0utTW31Fa9GasRRqo2Ik2XG+kdXv31e1pRrItzddXaUVpWroUDbdYbr9+t6q/HMo2BJJkkHR71VW1Fmq+100TZoZFu0uj4eF6Waeg3X7gsSXqiNJy2e3R/QV+7tihJ+sLZWT22fyi9160/tv51ml+Ov9qIVGtHKtdCzVTqmp6rp0kySao3Y52ZWdRMZaVstVNTQzpzraq4k6Rlz707l/5+d7m//rH64+m31XUEAAAAAAAAAAB7F4kydY9klKR2lCiMV5I1UdyRJIVRnNbpr9/TjpK03nr9bjTeZss368hYNyl2Ya6uS3M1FVxbw7nuDi7PsfTQmK+3r1dlm4ZOTQ2l7U7uH9K5m0u6stDQ+dmanly+92x1bP3r1K83724ia+0corijaiNaU95zaqqoRhjr/OxSWvbq9EL6e67WXjNWfzzrxQoAAAAAAAAAAHA7JMqk9A6zjG3IsYy0vLeryrGttE5//Z6MbaT11ut3o/E2W75ZvaMLL87VdGm+roMjORnGynwe3d9Njj08URjYjXZysqAwTvR/f/miJOnJvt1m/bH1r1O/3rwLWVt+Zu0cbKt7bOJGHj/QjevNq4tp2dszVR3fl5chaX6ptWas/njWixUAAAAAAAAAAOB2SJSpe3dZ4DtybUtj+YxG8xnlPEtZx1Yha6sUZAfu1urV73FtS4dHc/IzK8mgwF//nrT12t+u/maVgqxMQ7o4V9el+boOjw7e1XVyf0GS0vvJeh5bTqD9xvOXVMw6On14MFHWi61/nUZ8R4Yh5VxTSUfyMqYmizmVRnM6OrYybs6zdHJySJPFlbJqM9RstZkekXhsX14Z29QbVyqSujvQvnZtUY9NFTTiZ7TUihTGsQwjWTee9WIFAAAAAAAAAAC4nY23+ewxpSCnYjbUZNHVw5N5hVGijqQRP7Nu4qVXvxXGch1LBc9RtTn4vJnxNlt/M1zb0v5iVudna7pcbuivPzox8P7RyW5CbPWOsaNjvjK2qUoj1Mcfm+jedbbBXPrX6cpCQ4uNUIYpNdsdTZfry+8dnVxoqBHFGi+4A0my6XJ94F6xwHdUCnI6OVlId5R96dysGmFHx/cVdPbGkmZrLU0NZ2UZphxzcDfbdqwjAAAAAAAAAADYG0iU9entUrrT+nfb/l44MpbT8+fn1Y46OjgyuKPsG4+N6oc//JA++cTkQLltmXp4Iq83rizq2WOjt42tV24apoZzblperoUqZsNu28m1bavNcCBJ1t/m1NSQ/vD1GS022nrxQvd+suMTeY1f8vTK9IJsy0x37PWPc7tYAQAAAAAAAAAANsLRiw+Yw6O+Zhaby78HE2WeY+knP/mohnOZNe1OLu82e/b42KbGaYXxlspv1+bUVFGVRqiLczW9c3NJWcfSgeGsRvyMKvVQjVa4qb4AAAAAAAAAAAA2i0TZA+ZIX3Ls0KodZbfyqfdP6VPvn9KJ8fym6ruOtaXy27U5NdVN1J27WdM716s6Pp6XaRjaX/SUSKo0BxNj/+xz7+jP3rq+qVgBAAAAAAAAAADW88AkyqrNULPVpqrNcN3ne9HnbnB41JckWaahqeHsmvcbzelbTuzTP/++p2QYxi3r9RQ8R4E/eNxh4N/6CMRbtTk5OSTTkM7MLOnCXC1N2JWCbrJvsbESh2Ek+vXnL+mdG0sbjgUAAAAAAAAAAHA7D8QdZdPlenr3lSGpHXfkWCs5wMB30oTLnfR5p33shCPLibKpYW9gDaTNz2mz9UpBTsVsqFYYy3WsTd0TtlGbbMbS8fG8/r9XriiME33jsVHtL7py7WFJUtRJus+OpRcuzEuSnjo0fNvxAAAAAAAAAAAANrLrd5RVm+FAUqcVxbo4V1etHaVl5Vq4pV1hq/u8kz52Su+4xdXHLm52Tlude8FzNFbwNpUku12bU1NFXat071f70EOjGit4OjZekGMZml1qpW1evrQgyzT0ZKm46TEBAAAAAAAAAABW2/WJslY4eHdVO0okSWEU37LeVvq8kz52SjZj6cR4XqemBpNIm53TTs69d09ZkHNUCrrHRvaOkLxcbqT1Xr60oJOTBeUyD8SGSAAAAAAAAAAAsEN2fabBdayB54zdvWPLsa1b1ttKn3fSx0767R99Vu4m57+6fCfn3kvuPVEaTu9Kk6RSkNXl+bokKe4keuXygv7mU1PbHg8AAAAAAAAAAHiw7fodZQXPUeCvHOHn2pYOj+bk9+02Cnxny0cD9vd5J33spILnKGOba8o2M6ednPtjU0OyTUOnV909djDIabrcTZSdvbGkpVak04eCbY8HAAAAAAAAAAA82HbljrJqM1QrjOU6lgqeo2LWUdJJ1JHk2oZs01TU6cg2zbTOVpWCnIrZwXF2i9Xr09M/p976VJvhmrntxNyrzVBhFOuXf/BpPbTPH4jr4EhOs0tt1duRXr5UliQ9PJHXbLW5674NAAAAAAAAAAC4f+y6RFk77ujCbHd3kbH87FimDElztZbCTqKJgiepuxNqbPn3nSh4u2cXWc90ua5yLUyfA99RKcilzwXPUaUR3rJOr957Nffpcl0LtVBztZbK9Uhfv1HTRCFO4+rdVzZdbuilS2UNebY6HelapbVh/AAAAAAAAAAAALez645ejOMk/d2KYl2cq6vWjtSKYs3XQlUbkWrtSJJUroWqNsONunrgVJuDCTBp7Rpsps57qRdP7/tJSr9hL66DI90k2HS5rhcvlvXwRGHgDrO99p0BAAAAAAAAAMC9sesSZf3aUTdpFkZx+rv33NMK4zXtHlQbzbW/fDN13ku9cfu/n7TyDVthrIPLu8XeurqoczdrOjlZ2LAfAAAAAAAAAACAzdp1Ry/2y9jdXUWObckxV3YYObaV/nYda027B9VGc+0v30yd91Jv3N637Ol9Q9exlHdteY6pz752TZL0yOTQhv0AAAAAAAAAAABs1q7bUWZZKwkV17Z0eDQnP2PLtS2N+I4KWVt+ppv/C/zdd8fY3Sh4jgJ/cL6r12Azdd5LvXh6309S+g17cRmGoVKQ05mZqgxD+sDRQO0o1lIzUjuK99x3BgAAAAAAAAAA98au21GWsUwdGcupFcZyHUsFz1G1GaoVxjo81j2ir//dXlMKcipmw1uuwWbqvJd68UwWXUWdjmzTXBPXwSCrszeW9PB4Qb5rq1wPFXViGR12kgEAAAAAAAAAgDuz6xJlUncX0updUquf97LV63Gndd5Lt4vn4Eg3Cfr4gSGVa2F31+DyzsFyLVQxG95X8wEAAAAAAAAAAPe/XXf0Ivamg0E3Ufbo/rX3k0ndXYQAAAAAAAAAAABbQaIMu8LjB4rKWKaePTa67nvX4QhGAAAAAAAAAACwNbvy6EXsPd94bFQv/w8fk+/ami7XVa6F6bvAv7+OkQQAAAAAAAAAALvDrkuUdZJEs9WmXMciOXKHqs1QrTC+b9dwo/h8t/vftRTkVMze33MAAAAAAAAAAAD3v12XKGuFHV2rtCR1dxKVlu+uwuastxvrflrDzcZX8NhFBgAAAAAAAAAA7s6uvqOsXAtVbYa3rwhJ3Z1a/Uko6f5aw/s9PgAAAAAAAAAA8GDZ1YkySWqF8U6HsGtstFb3yxre7/EBAAAAAAAAAIAHy65PlLmOtdMh7BobrdX9sob3e3wAAAAAAAAAAODBsqsTZYHPPVVS98jC2WrztkcUFjxHgT+4XvfTGm4mvs3OFQAAAAAAAAAA4HbsnQ5gq1zH1P6iK9ex7psEz06aLtcH7vUKfEelILdh/VKQUzEbqhXG9+Ua3iq+rc4VAAAAAAAAAADgVnZdosw0DI0VvJ0O475QbYYDiSNJKtdCFbPhLRNgBe/+2UW2nvXiu9O5AgAAAAAAAAAAbGRXH72417XCeEvlu9lemisAAAAAAAAAAHhvkCjbxVzH2lL5braX5goAAAAAAAAAAN4bJMp2sYLnKPAHjx0M/Pv7WMU7tZfmCgAAAAAAAAAA3hu77o4yDCoFORWzoVphLNexHqjEUbU5OK8Hea4AAAAAAAAAAOC9R6LsAVDwHrydVdPlusq1MH0O/G6i7EGcKwAAAAAAAAAA2BkcvYj7TrUZDiTJJKlcC1Vthhu0AAAAAAAAAAAA2DoSZbjvtMJ4S+UAAAAAAAAAAAB3gkQZ7juuY22pHAAAAAAAAAAA4E6QKMOOqjZDzVabA8cqFjxHgT94D1ngczcZAAAAAAAAAAC4t+ydDgB713S5PnAXWeA7KgU5SVIpyKmYDdUKY7mORZIMAAAAAAAAAADccyTKsCOqzXAgSSZJ5VqoYjZMk2IFj11kAAAAAAAAAABg+3D0InZEK4y3VA4AAAAAAAAAAHCvkSjDjnAda0vlAAAAAAAAAAAA9xqJMuyIguco8AePVQz8rR+1WG2Gmq02VW2Gt68MAAAAAAAAAADQhzvKsGNKQU7FbKhWGMt1rC0nyabL9YF7zgLfUSnI3eswAQAAAAAAAADAA4pEGXZUwdv6LjKpu5OsP0kmSeVaqGI2vKP+AAAAAAAAAADA3sPRi9iVWmG8pXIAAAAAAAAAAIDVSJRhV3Ida0vlAAAAAAAAAAAAq21roswwjE8YhvG2YRhnDcP4iXXeFw3D+H3DMF41DONNwzB+aDvjwYOj4DkK/MEjFgP/zo5xBAAAAAAAAAAAe9O23VFmGIYl6eckfUzStKQXDMP4vSRJ3uqr9o8kvZUkyXcZhrFP0tuGYfxakiTt7YoLD45SkFMxG6oVxnIdiyQZAAAAAAAAAADYkm1LlEn6gKSzSZK8K0mGYfw/kj4lqT9RlkgqGIZhSMpLmpcUbWNMeMAUPHaRAQAAAAAAAACAO7OdRy8ekHS573l6uazfz0p6VNJVSa9L+sdJknS2MSYAAAAAAAAAAABA0vYmyox1ypJVz98u6RVJU5LeL+lnDcMYWtORYfywYRhfNQzjqzdv3rz3kQIAAAAAAAAAAGDP2c5E2bSkg33PJXV3jvX7IUm/nXSdlXRe0snVHSVJ8otJkjyTJMkz+/bt27aAAQAAAAAAAAAAsHdsZ6LsBUknDMM4ahhGRtL3Sfq9VXUuSfqoJBmGMSHpEUnvbmNMAAAAAAAAAAAAgCTJ3q6OkySJDMP4MUl/IsmS9K+SJHnTMIx/uPz+FyT9tKR/bRjG6+oe1fjfJEkyu10xAQAAAAAAAAAAAD3bliiTpCRJ/lDSH64q+4W+31clfXw7YwAAAAAAAAAAAADWs51HLwIAAAAAAAAAAAD3LRJlAAAAAAAAAAAA2JNIlAEAAAAAAAAAAGBPIlEGAAAAAAAAAACAPYlEGQAAAAAAAAAAAPYkEmUAAAAAAAAAAADYk0iUAQAAAAAAAAAAYE8iUQYAAAAAAAAAAIA9iUQZAAAAAAAAAAAA9iQSZQAAAAAAAAAAANiTSJQBAAAAAAAAAABgTyJRBgAAAAAAAAAAgD2JRBkAAAAAAAAAAAD2JBJlAAAAAAAAAAAA2JPsnQ4A2KxqM1QrjOU6lgqes9PhAAAAAAAAAACAXY5EGXaF6XJd5VqYPge+o1KQ28GIAAAAAAAAAADAbsfRi7jvVZvhQJJMksq1UNVmuEELAAAAAAAAAACA2yNRhvteK4y3VA4AAAAAAAAAALAZJMpw33Mda0vlAAAAAAAAAAAAm0GiDGtUm6Fmq8375mjDguco8J2BssB3VPCcDVoAAAAAAAAAAADcnr3TAeD+Ml2uD9wHFviOSkFuByPqKgU5FbOhWmEs17FIkgEAAAAAAAAAgLtGogypajMcSJJJUrkWqpgN74vEVMFjFxkAAAAAAAAAALh3OHoRqVYYb6kcAAAAAAAAAABgNyNRhpTrWFsqBwAAAAAAAAAA2M1IlCFV8BwF/uDRhoHPcYcAAAAAAAAAAODBxB1lGFAKcipmQ7XCWK5jkSQDAAAAAAAAAAAPLBJlWKPgsYsMAAAAAAAAwP/f3r0He17X9x1/vdldQQlGjGgsF11TvJCMom5RK21QcwEnllxH0HhNZ2Mj9ZbphDo2aWubSdqxtjYgRUu9REPsiJG2CnYcNBpjZFGLIqArqKxQUSEJJYJc3v3j/BZ/bs85e47d7+/s7ufxmPnN+X0v58f78MfHnzzn+/0CHPzcehEAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJA2b/QATOu2O+7KnXfdk0O3bMoRh23Z6HEAAAAAAAD2G0LZQWzXrX+TW2+/677tIw/fkmOOfMAGTgQAAAAAALD/cOvFg9Rtd9z1fZEsSW69/a7cdsddK/wGAAAAAADAWISyg9Sdd92zrv0AAAAAAACjEcoOUodu2bSu/QAAAAAAAKMRyg5SRxy2JUcevuX79h15+JYccdiWFX4DAAAAAABgLJs3egCmc8yRD8gP3/+u3HnXPTl0yyaRDAAAAAAAYI5QdpA74jBXkQEAAAAAACzHrRcBAAAAAAAYklAGAAAAAADAkIQyAAAAAAAAhiSUAQAAAAAAMCShDAAAAAAAgCEJZQAAAAAAAAxJKAMAAAAAAGBIQhkAAAAAAABDEsoAAAAAAAAYklAGAAAAAADAkIQyAAAAAAAAhiSUAQAAAAAAMCShDAAAAAAAgCEJZQAAAAAAAAxJKAMAAAAAAGBIQhkAAAAAAABDEsoAAAAAAAAYklAGAAAAAADAkIQyAAAAAAAAhiSUAQAAAAAAMCShDAAAAAAAgCEJZQAAAAAAAAxJKAMAAAAAAGBIQhkAAAAAAABDEsoAAAAAAAAYklAGAAAAAADAkIQyAAAAAAAAhiSUAQAAAAAAMCShDAAAAAAAgCEJZQAAAAAAAAxJKAMAAAAAAGBIQhkAAAAAAABDEsoAAAAAAAAYklAGAAAAAADAkIQyAAAAAAAAhjRpKKuqU6vq2qraWVVnr3DOKVX12aq6qqo+OuU8AAAAAAAAsNvmqT64qjYlOSfJTyfZleTyqrq4u78wd86Dkpyb5NTu/lpVPXSqeQAAAAAAAGDelFeUnZRkZ3df193fTXJhktP3OOd5SS7q7q8lSXffPOE8AAAAAAAAcJ8pQ9nRSW6Y29412zfv0UmOrKqPVNUVVfXCCecBAAAAAACA+0x268Uktcy+Xuaf/+Qkz0py/yR/XlWf7O4vft8HVW1Psj1JjjvuuAlGBQAAAAAAYDRTXlG2K8mxc9vHJLlxmXMu6e7bu/tbSf40yRP2/KDuPr+7t3X3tqOOOmqygQEAAAAAABjHlKHs8iTHV9XWqrpfkjOSXLzHOe9P8veqanNVPSDJU5JcPeFMAAAAAAAAkGTCWy92991VdVaSS5NsSnJBd19VVS+bHT+vu6+uqkuSXJnk3iRv7e7PTzUTAAAAAAAA7Fbdez42bP+2bdu23rFjx0aPAQAAAAAAwIJV1RXdvW1ffd6Ut14EAAAAAACA/ZZQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMaa+hrKp+YhGDAAAAAAAAwCKt5Yqy86rqU1X1G1X1oMknAgAAAAAAgAXYayjr7pOTPD/JsUl2VNW7q+qnJ58MAAAAAAAAJrSmZ5R195eSvC7JbyX5ySRvqqprquoXpxwOAAAAAAAAprKWZ5Q9vqremOTqJM9M8pzuftzs/Rsnng8AAAAAAAAmsXkN5/xBkrckeW13f2f3zu6+sapeN9lkAAAAAAAAMKG13Hrxou5+53wkq6pXJkl3v3OyyQAAAAAAAGBCawllL1xm34v38RwAAAAAAACwUCveerGqzkzyvCRbq+riuUNHJPn21IMBAAAAAADAlFZ7RtknktyU5CFJ3jC3/7YkV045FAAAAAAAAExtxVDW3V9N8tUkT1vcOAAAAAAAALAYq9168ePdfXJV3Zak5w8l6e5+4OTTAQAAAAAAwERWu6Ls5NnPIxY3DgAAAAAAACzGaleUPXi1X+zuW/b9OAAAAAAAALAYK4ayJFdk6ZaLtcyxTvKoSSYCAAAAAACABVjt1otbFzkIAAAAAAAALNJqt158bHdfU1VPWu54d396urEAAAAAAABgWqvdevE1SbYnecMyxzrJMyeZCAAAAAAAABZgtVsvbp/9fMbixgEAAAAAAIDFWO2KsiRJVR2W5DeSnJylK8k+luS87r5j4tkAAAAAAABgMnsNZUnekeS2JP9xtn1mkncm+ZWphgIAAAAAAICprSWUPaa7nzC3fVlV/a+pBgIAAAAAAIBFOGQN53ymqp66e6OqnpLkz6YbCQAAAAAAAKa34hVlVfW5LD2TbEuSF1bV12bbj0jyhcWMBwAAAAAAANNY7daLP7ewKQAAAAAAAGDBVgxl3f3V+e2qemiSwyafCAAAAAAAABZgr88oq6p/UFVfSnJ9ko8m+UqSD048FwAAAAAAAExqr6EsyeuTPDXJF7t7a5JnJfmzSacCAAAAAACAia0llN3V3d9OckhVHdLdlyU5ceK5AAAAAAAAYFIrPqNszl9W1Q8l+ViSd1XVzUnunnYsAAAAAAAAmNZarig7Pcl3krwqySVJvpzkvtMWcwAAF7ZJREFUOVMOBQAAAAAAAFPb6xVl3X17Vf1okpOS3JLk0tmtGAEAAAAAAOCAtdcryqrqHyb5VJJfTPLLST5ZVS+dejAAAAAAAACY0lqeUfZPkjxx91VkVfUjST6R5IIpBwMAAAAAAIApreUZZbuS3Da3fVuSG6YZBwAAAAAAABZjxSvKquo1s7dfT/IXVfX+JJ3k9CzdihEAAAAAAAAOWKvdevGI2c8vz167vX+6cQAAAAAAAGAxVgxl3f0v5rer6oil3f1/Jp8KAAAAAAAAJrbXZ5RV1U9U1WeSfD7JVVV1RVX9+PSjAQAAAAAAwHT2GsqSnJ/kNd39iO5+RJLfTPKWaccCAAAAAACAaa0llB3e3Zft3ujujyQ5fLKJAAAAAAAAYAFWfEbZnOuq6p8leeds+1eTXD/dSAAAAAAAADC9tVxR9tIkRyW5aPZ6SJKXTDkUAAAAAAAATG3VK8qqalOS13b3KxY0DwAAAAAAACzEqleUdfc9SZ68oFkAAAAAAABgYdbyjLLPVNXFSf5rktt37+zuiyabCgAAAAAAACa2llD24CTfTvLMuX2dpeeVAQAAAAAAwAFpr6Gsu1+yiEEAAAAAAABgkVZ9RlmSVNWjquq/VdU3q+rmqnp/VW1dxHAAAAAAAAAwlb2GsiTvTvKeJA9P8rey9KyyC6ccCgAAAAAAAKa2llBW3f3O7r579vrDLD2jDAAAAAAAAA5Ye31GWZLLqursLF1F1kmem+R/VNWDk6S7b5lwPgAAAAAAAJjEWkLZc2c/f32P/S/NUjh71D6dCAAAAAAAABZgr6Gsu7cuYhAAAAAAAABYpLU8owwAAAAAAAAOOkIZAAAAAAAAQxLKAAAAAAAAGNJeQ1kt+dWq+u3Z9nFVddL0owEAAAAAAMB01nJF2blJnpbkzNn2bUnOmWwiAAAAAAAAWIDNazjnKd39pKr6TJJ0961Vdb+J5wIAAAAAAIBJreWKsruqalOSTpKqOirJvZNOBQAAAAAAABNbSyh7U5L3JXloVf3rJB9P8ruTTgUAAAAAAAAT2+utF7v7XVV1RZJnJakkP9/dV08+GQAAAAAAAExor6Gsqv5Dkj/u7nMWMA8AAAAAAAAsxFpuvfjpJK+rqp1V9W+ratvUQwEAAAAAAMDU9hrKuvvt3f3sJCcl+WKS36+qL00+GQAAAAAAAExoLVeU7fa3kzw2ySOTXDPJNAAAAAAAALAgew1lVbX7CrJ/meSqJE/u7udMPhkAAAAAAABMaPMazrk+ydO6+1tTDwMAAAAAAACLsmIoq6rHdvc1ST6V5LiqOm7+eHd/eurhAAAAAAAAYCqrXVH2miTbk7xhmWOd5JmTTAQAAAAAAAALsGIo6+7ts7endfcd88eq6rBJpwIAAAAAAICJHbKGcz6xxn0AAAAAAABwwFjtGWU/muToJPevqicmqdmhByZ5wAJmAwAAAAAAgMms9oyyn03y4iTHZOk5ZbtD2V8nee20YwEAAAAAAMC0VntG2duTvL2qfqm737vAmQAAAAAAAGBya3lG2ZOr6kG7N6rqyKr6VxPOBAAAAAAAAJNbSyg7rbv/cvdGd9+a5NnTjQQAAAAAAADTW0so21RVh+7eqKr7Jzl0lfMBAAAAAABgv7fiM8rm/GGSD1fVf0nSSV6a5B2TTgUAAAAAAAAT22so6+5/U1VXJvmpJJXk9d196eSTAQAAAAAAwITWckVZuvuSJJckSVU9varO6e6XTzoZAAAAAAAATGhNoayqTkxyZpLnJrk+yUVTDgUAAAAAAABTWzGUVdWjk5yRpUD27SR/nKS6+xkLmg0AAAAAAAAms9oVZdck+ViS53T3ziSpqlcvZCoAAAAAAACY2CGrHPulJP87yWVV9ZaqelaSWsxYAAAAAAAAMK0VQ1l3v6+7n5vksUk+kuTVSR5WVW+uqp9Z0HwAAAAAAAAwidWuKEuSdPft3f2u7v65JMck+WySsyefDAAAAAAAACa011A2r7tv6e7/1N3PnGogAAAAAAAAWIR1hTIAAAAAAAA4WAhlAAAAAAAADEkoAwAAAAAAYEhCGQAAAAAAAEOaNJRV1alVdW1V7ayqs1c57+9U1T1V9ctTzgMAAAAAAAC7TRbKqmpTknOSnJbkhCRnVtUJK5z3+0kunWoWAAAAAAAA2NOUV5SdlGRnd1/X3d9NcmGS05c57x8neW+SmyecBQAAAAAAAL7PlKHs6CQ3zG3vmu27T1UdneQXkpw34RwAAAAAAADw/5gylNUy+3qP7X+f5Le6+55VP6hqe1XtqKod3/zmN/fZgAAAAAAAAIxr84SfvSvJsXPbxyS5cY9ztiW5sKqS5CFJnl1Vd3f3n8yf1N3nJzk/SbZt27ZnbAMAAAAAAIB1mzKUXZ7k+KramuTrSc5I8rz5E7p76+73VfW2JP99z0gGAAAAAAAAU5gslHX33VV1VpJLk2xKckF3X1VVL5sd91wyAAAAAAAANsyUV5Sluz+Q5AN77Fs2kHX3i6ecBQAAAAAAAOYdstEDAAAAAAAAwEYQygAAAAAAABiSUAYAAAAAAMCQhDIAAAAAAACGJJQBAAAAAAAwJKEMAAAAAACAIQllAAAAAAAADEkoAwAAAAAAYEibN3qAdbv22uSUUzZ6CgAAAAAAAA5wrigDAAAAAABgSAfeFWWPeUzykY9s9BQAAAAAAAAsWtU+/ThXlAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADCkSUNZVZ1aVddW1c6qOnuZ48+vqitnr09U1ROmnAcAAAAAAAB2myyUVdWmJOckOS3JCUnOrKoT9jjt+iQ/2d2PT/L6JOdPNQ8AAAAAAADMm/KKspOS7Ozu67r7u0kuTHL6/And/YnuvnW2+ckkx0w4DwAAAAAAANxnylB2dJIb5rZ3zfat5NeSfHDCeQAAAAAAAOA+myf87FpmXy97YtUzshTKTl7h+PYk25PkuOOO21fzAQAAAAAAMLApryjbleTYue1jkty450lV9fgkb01yend/e7kP6u7zu3tbd2876qijJhkWAAAAAACAsUwZyi5PcnxVba2q+yU5I8nF8ydU1XFJLkrygu7+4oSzAAAAAAAAwPeZ7NaL3X13VZ2V5NIkm5Jc0N1XVdXLZsfPS/LbSX4kyblVlSR3d/e2qWYCAAAAAACA3ap72ceG7be2bdvWO3bs2OgxAAAAAAAAWLCqumJfXnQ15a0XAQAAAAAAYL8llAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMSSgDAAAAAABgSEIZAAAAAAAAQxLKAAAAAAAAGJJQBgAAAAAAwJCEMgAAAAAAAIYklAEAAAAAADAkoQwAAAAAAIAhCWUAAAAAAAAMadJQVlWnVtW1VbWzqs5e5nhV1Ztmx6+sqidNOQ8AAAAAAADsNlkoq6pNSc5JclqSE5KcWVUn7HHaaUmOn722J3nzVPMAAAAAAADAvCmvKDspyc7uvq67v5vkwiSn73HO6Une0Us+meRBVfXwCWcCAAAAAACAJNOGsqOT3DC3vWu2b73nAAAAAAAAwD63ecLPrmX29Q9wTqpqe5ZuzZgkd1bV5/8/ZwMYzUOSfGujhwA4wFg7AdbP2gmwPtZNgPV7zL78sClD2a4kx85tH5Pkxh/gnHT3+UnOT5Kq2tHd2/btqAAHN2snwPpZOwHWz9oJsD7WTYD1q6od+/Lzprz14uVJjq+qrVV1vyRnJLl4j3MuTvLCWvLUJH/V3TdNOBMAAAAAAAAkmfCKsu6+u6rOSnJpkk1JLujuq6rqZbPj5yX5QJJnJ9mZ5G+SvGSqeQAAAAAAAGDelLdeTHd/IEsxbH7feXPvO8nL1/mx5++D0QBGY+0EWD9rJ8D6WTsB1se6CbB++3TtrKVWBQAAAAAAAGOZ8hllAAAAAAAAsN86oEJZVZ1aVddW1c6qOnuj5wHYX1TVV6rqc1X12araMdv34Kr6n1X1pdnPI+fO/6eztfTaqvrZjZscYHGq6oKqurmqPj+3b91rZVU9ebbm7qyqN1VVLfpvAViUFdbOf15VX5999/xsVT177pi1ExhaVR1bVZdV1dVVdVVVvXK23/dOgBWssnYu5HvnARPKqmpTknOSnJbkhCRnVtUJGzsVwH7lGd19Yndvm22fneTD3X18kg/PtjNbO89I8uNJTk1y7myNBTjYvS1L6968H2StfHOS7UmOn732/EyAg8nbsvw698bZd88TZ88nt3YCLLk7yW929+OSPDXJy2fro++dACtbae1MFvC984AJZUlOSrKzu6/r7u8muTDJ6Rs8E8D+7PQkb5+9f3uSn5/bf2F339nd1yfZmaU1FuCg1t1/muSWPXava62sqocneWB3/3kvPez3HXO/A3DQWWHtXIm1Exhed9/U3Z+evb8tydVJjo7vnQArWmXtXMk+XTsPpFB2dJIb5rZ3ZfV/UQAj6SQfqqorqmr7bN/DuvumZOl/bJI8dLbfegrwPetdK4+evd9zP8BozqqqK2e3Ztx9+zBrJ8Ccqnpkkicm+Yv43gmwJnusnckCvnceSKFsuftI9sKnANg/Pb27n5Sl29O+vKr+/irnWk8B9m6ltdIaCrB0O5sfS3JikpuSvGG239oJMFNVP5TkvUle1d1/vdqpy+yzdgJDWmbtXMj3zgMplO1Kcuzc9jFJbtygWQD2K9194+znzUnel6VbKX5jdrlxZj9vnp1uPQX4nvWulbtm7/fcDzCM7v5Gd9/T3fcmeUu+dxtvaydAkqrakqX/0Puu7r5ottv3ToBVLLd2Lup754EUyi5PcnxVba2q+2XpQW0Xb/BMABuuqg6vqiN2v0/yM0k+n6U18kWz016U5P2z9xcnOaOqDq2qrVl6qOWnFjs1wH5jXWvl7DY5t1XVU6uqkrxw7ncAhrD7P/TO/EKWvnsm1k6AzNa5/5zk6u7+d3OHfO8EWMFKa+eivndu3kd/x+S6++6qOivJpUk2Jbmgu6/a4LEA9gcPS/K+pbU/m5O8u7svqarLk7ynqn4tydeS/EqSdPdVVfWeJF9IcneSl3f3PRszOsDiVNUfJTklyUOqaleS30nye1n/WvmPkrwtyf2TfHD2AjgorbB2nlJVJ2bpNjZfSfLribUTYObpSV6Q5HNV9dnZvtfG906A1ay0dp65iO+d1e3WtgAAAAAAAIznQLr1IgAAAAAAAOwzQhkAAAAAAABDEsoAAAAAAAAYklAGAAAAAADAkIQyAAAAAAAAhiSUAQAAbJCqekVVXV1V79roWQAAAEZU3b3RMwAAAAypqq5Jclp3Xz+3b3N3372BYwEAAAzDFWUAAAAboKrOS/KoJBdX1V9V1flV9aEk76iqR1bVx6rq07PX3539zilV9dGqek9VfbGqfq+qnl9Vn6qqz1XVj83OO6qq3ltVl89eT9/APxUAAGC/5YoyAACADVJVX0myLclZSZ6T5OTu/k5VPSDJvd19R1Udn+SPuntbVZ2S5E+SPC7JLUmuS/LW7v6dqnplkq3d/aqqeneSc7v741V1XJJLu/txi/8LAQAA9m+bN3oAAAAAkiQXd/d3Zu+3JPmDqjoxyT1JHj133uXdfVOSVNWXk3xotv9zSZ4xe/9TSU6oqt2/88CqOqK7b5vyDwAAADjQCGUAAAD7h9vn3r86yTeSPCFLt8y/Y+7YnXPv753bvjff+/94hyR52lx4AwAAYBmeUQYAALD/+eEkN3X3vUlekGTTOn//Q1m6nWOSZHZlGgAAAHsQygAAAPY/5yZ5UVV9Mku3Xbx9L+fv6RVJtlXVlVX1hSQv29cDAgAAHAyquzd6BgAAAAAAAFg4V5QBAAAAAAAwJKEMAAAAAACAIQllAAAAAAAADEkoAwAAAAAAYEhCGQAAAAAAAEMSygAAAAAAABiSUAYAAAAAAMCQhDIAAAAAAACG9H8BP0xOgbROQ+IAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 2160x720 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "traj_viz(pdb_id, path_to_gpcrdb_files, dir_path, out_path, one_mod_df, bi_mod_df, model, frames_dir_path, 2500, new_seq_aligned_to_GPCRdb_seq_database, canonical_residues_dict)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
